NXP support lptmr tickless function

yongxu-wang15 opened the pull request:
1. added lptmr tickless function
2. verified in imx95 m7 core which enabled PM and tickless
![Uploading image.png…]()

yongxu-wang15 committed:
dts: arm: nxp: imx95_m7: add lptmr1 node

add lptmr1 node into imx95 m7 dts which locate on Always On
domain and can continue work when whole soc enter system sleep mode

Signed-off-by: Yongxu Wang <yongxu.wang@nxp.com>

yongxu-wang15 committed:
board: nxp: imx95_evk: set lptmr1 interface

enable lptmr1 interface as default for system timer

Signed-off-by: Yongxu Wang <yongxu.wang@nxp.com>

yongxu-wang15 committed:
test: counter_basic_api: add imx95 evk m7 support

Enable lptmr interface to support counter_basic_api test case

Signed-off-by: Yongxu Wang <yongxu.wang@nxp.com>

yongxu-wang15 committed:
soc: nxp: imx95: use lptmr as system tick

use lptmr timer as system tick source

Signed-off-by: Yongxu Wang <yongxu.wang@nxp.com>

yongxu-wang15 committed:
drivers: mcux_lptmr_timer: support tickless function

When tickless is enabled, it is necessary to ensure
that lptmr can be set for a period of time instead of generating
an interrupt at one tick.

LPTMR hardware does not synchronize writes to the
compare register (CMR)
If CMR is changed exactly when the counter increments,
the hardware may compare part of the counter against
the old CMR value and part against the new one, leading to TCF
being set incorrectly on some SoCs.

Reorder the update sequence to:
- disable the LPTMR interrupt,
- clear TCF,
- write the new CMR,
- re-enable the interrupt.

This provides a safe workaround for tickless mode case.

Signed-off-by: Yongxu Wang <yongxu.wang@nxp.com>

yongxu-wang15 commented:
Updated to fixed dts conflict with newly commit modification.

Hi, this PR is aim to support nxp lptmr tickless function, and supported it in imx95 m7, can you help review it when you time free?

Thanks

decsny commented:
@mmahadevan108 please help review this

yongxu-wang15 commented:
Hi @mmahadevan108 , I have added disable/start code logic in LPTMR_SetTimerPeriod to avoid the hardware limitations, tested passed on imx95 m7 power_mgmt_soc test case, the time sleep and tick record is right in my side, please help reivew it when you time free, thanks

henrikbrixandersen reviewed (changes requested):
Stopping/starting the timer when setting a new timeout _will_ affect the ticks (as the time all the instructions between stopping and starting will not be counted). I don't think this is an acceptable solution.

yongxu-wang15 commented:
Hi @henrikbrixandersen I fully understand your concerns. After reconfirming with the design team, we need to ensure that there is sufficient time when updating the CMP. We can choose to reset(but this will cause the time to go back instead of monotonically increasing. Even through software, it is difficult to guarantee that there are no corner cases). In the latest commit, I added a protection mechanism. When the distance is very close, It will postpone a windo to prevent it from being triggered. Of course, this should depend on different SoCs, please help review it again when you time free, thanks a lot

yongxu-wang15 commented:
Hi @henrikbrixandersen, Has there been any progress on this issue? thanks

teburd commented:
The design of kernel timers is time never stops (free running counter) that you may set a compare to catch sometime in the future.

Stopping/Starting the underlying timer will *surely* introduce timing drift and jitter, something that tends to surprise and upset people when they see it. My opinion is that if your hardware needs to stop and start to set the compare, this hardware isn't a good source for kernel timers without at least some build warnings informing people building with it that yes, this is what they really want.

I'm not going to stop a vendor from supplying a driver that introduces drift/jitter. But I do feel like I can say we should warn about it. Unless you can perhaps show in the PR the external tool test from https://github.com/zephyrproject-rtos/zephyr/tree/main/tests/kernel/timer/timer_behavior with correct timer behavior with this driver with the start/stop behavior? Maybe its ok, I don't know!

Holt-Sun commented on soc/nxp/imx/imx9/imx95/Kconfig.defconfig.mimx95.m7:
```diff
@@ -42,19 +42,33 @@ config 3RD_LEVEL_INTERRUPTS
+	default y
+
+config SYS_CLOCK_TICKS_PER_SEC
+	default 1000
```
Consider to set default to 1024.
For default 1000, since SYS_CLOCK_HW_CYCLES_PER_SEC = 32768, then CYCLES_PER_TICK = 32768 / 1000 = 32 (fractional part is dropped), so one OS “tick” is 32 hardware cycles, than tick_rate = tick_rate = 32768 / 32 = 1024 ticks/s, but the configuration says "1000 ticks/s", there is 2.4% drift.

For a 32768 Hz clock, Zephyr typically uses tick rates that divide 32768 exactly (1024, 512, 256, 128, …), in this way, there is no rafe shift.

zejiang0jason commented on drivers/timer/Kconfig.mcux_lptmr:
```diff
@@ -13,3 +14,23 @@ config MCUX_LPTMR_TIMER
+	  If enabled, the driver will use SoC-specific default values.
+
+config MCUX_LPTMR_TIMER_SAFETY_WINDOW_CYCLES
+	int "MCUX LPTMR timer safety window cycles"
```
How can user choose a proper value, is there a guidance?

zejiang0jason commented on drivers/timer/mcux_lptmr_timer.c:
```diff
@@ -32,16 +35,94 @@ BUILD_ASSERT(DT_NUM_INST_STATUS_OKAY(DT_DRV_COMPAT) == 1,
+		next += CYCLES_PER_TICK;
+	}
+
+	next += announced_cycles;
```
could this `next` overflow? which means in function `lptmr_set_safe_immediate`, target_cycles < hw_counter

zejiang0jason commented on soc/nxp/imx/imx9/imx95/Kconfig.defconfig.mimx95.m7:
```diff
@@ -42,19 +42,33 @@ config 3RD_LEVEL_INTERRUPTS
+	default y
+
+config SYS_CLOCK_TICKS_PER_SEC
+	default 1024
```
can be
```
default 1024 if MCUX_LPTMR_TIMER
default 1000 if CORTEX_M_SYSTICK
```

Albort12138 commented on drivers/timer/mcux_lptmr_timer.c:
```diff
@@ -83,7 +186,13 @@ static int sys_clock_driver_init(void)
+		"tickless mode, which may introduce timing drift and jitter");
+#else
 	config.enableFreeRunning = false;
+#endif
```
What is the default compare value before invoking sys_clock_set_timeout?

yongxu-wang15 replied:
it will be CYCLES_PER_TICK as default in sys_clock_driver_init

yongxu-wang15 commented on drivers/timer/mcux_lptmr_timer.c:
the lptmr as system timer also use counter/nxp,lptmr.yaml now, its same with counter lptmr, in my test case, not enabled CONFIG_COUNTER, so not have conflict in this

yeaissa commented on drivers/timer/mcux_lptmr_timer.c:
The timer driver shall select TICKLESS_CAPABLE Kconfig, to mention that it's capable of supporting tickless operations

yeaissa commented on drivers/timer/mcux_lptmr_timer.c:
Copyright to be updated?

Holt-Sun commented on soc/nxp/imx/imx9/imx95/Kconfig.defconfig.mimx95.m7:
```diff
@@ -42,14 +42,29 @@ config 3RD_LEVEL_INTERRUPTS
+
 config SYS_CLOCK_HW_CYCLES_PER_SEC
-	default 800000000
+	default 32768 if MCUX_LPTMR_TIMER
```
The `clock_frequency` is defined in [dts/bindings/timer/nxp,lptmr-timer.yaml](https://github.com/nxp-upstream/zephyr/blob/f015add92d033cfb384a68bec5e5f4c47351ed7e/dts/bindings/timer/nxp%2Clptmr-timer.yaml#L17) and Instantiated in the [dts/arm/nxp/imx/nxp_imx95_m7.dtsi](https://github.com/nxp-upstream/zephyr/blob/f015add92d033cfb384a68bec5e5f4c47351ed7e/dts/arm/nxp/imx/nxp_imx95_m7.dtsi#L615), then the value can be got from device tree.

yongxu-wang15 committed:
test: ram_context_for_isr: remove imx95 flash target

fix twister build issue, remove imx95 flash target from this test case

Signed-off-by: Yongxu Wang <yongxu.wang@nxp.com>

yongxu-wang15 commented:
Hi @henrikbrixandersen can you help reivew this pr again? it use new way to update lptmr cmp with internal design team, thanks

Holt-Sun commented:
I tried locally with frdm_mcxc444, works fine.

yongxu-wang15 committed:
drivers: timer: mcux_lptmr: add prescaler bypass support

Add support for prescaler bypass configuration via devicetree
property. This aligns the timer driver with the counter driver
implementation.

Signed-off-by: Yongxu Wang <yongxu.wang@nxp.com>

yongxu-wang15 commented on drivers/timer/mcux_lptmr_timer.c:
```diff
@@ -1,11 +1,11 @@
  */
 
-#define DT_DRV_COMPAT nxp_lptmr
+#define DT_DRV_COMPAT nxp_lptmr_hw_timer
```
its better to split compatible from counter from my understand, but it need some time to update soc/dts among with different soc, not plan to add it in this pr, created one link to track it: https://github.com/zephyrproject-rtos/zephyr/issues/104445

yongxu-wang15 commented on drivers/timer/mcux_lptmr_timer.c:
```diff
@@ -83,7 +175,11 @@ static int sys_clock_driver_init(void)
 	config.enableFreeRunning = false;
+#endif
 	config.prescalerClockSource = LPTMR_CLK_SOURCE;
 	config.bypassPrescaler = (LPTMR_PRESCALER == 0);
```
sure, done, thanks

Holt-Sun commented:
Hi @nordic-krch, please review this PR, thanks.
